com.unity.shadergraph/Editor/Drawing/Controllers/ShaderInputViewController.cs

```diff
@@ -0,0 +1,304 @@
+            graphData.owner.RegisterCompleteObjectUndo("Reset Reference Name");
+            ShaderInputReference.overrideReferenceName = null;
+        }
+
+        public Action<GraphData> modifyGraphDataAction =>  ResetReferenceName;
+
+        // Reference to the shader input being modified
+        internal ShaderInput ShaderInputReference { get; set; }
+    }
+
+    class DeleteShaderInputAction : IGraphDataAction
+    {
+        void DeleteShaderInput(GraphData graphData)
+        {
+            Assert.IsNotNull(graphData, "GraphData is null while carrying out DeleteShaderInputAction");
+            Assert.IsNotNull(shaderInputsToDelete, "ShaderInputReference is null while carrying out DeleteShaderInputAction");
+
+            // This is called by MaterialGraphView currently, no need to repeat it here, though ideally it would live here
+            //graphData.owner.RegisterCompleteObjectUndo("Delete Graph Input(s)");
+
+            foreach (var shaderInput in shaderInputsToDelete)
+            {
+                graphData.RemoveGraphInput(shaderInput);
+            }
+        }
+
+        public Action<GraphData> modifyGraphDataAction =>  DeleteShaderInput;
+
+        // Reference to the shader input(s) being deleted
+        internal IList<ShaderInput> shaderInputsToDelete { get; set; } = new List<ShaderInput>();
+    }
+
+    class ShaderInputViewController : SGViewController<ShaderInput, ShaderInputViewModel>
+    {
+        // Exposed for PropertyView
+        internal GraphData graphData => DataStore.State;
+
+        internal ShaderInputViewController(ShaderInput shaderInput, ShaderInputViewModel inViewModel, GraphDataStore graphDataStore)
+            : base(shaderInput, inViewModel, graphDataStore)
+        {
+            InitializeViewModel();
+
+            m_BlackboardPropertyView = new BlackboardPropertyView(ViewModel);
+            m_BlackboardPropertyView.controller = this;
+            m_BlackboardPropertyView.RegisterCallback<AttachToPanelEvent>((evt) => ViewModel.updateSelectionStateAction(evt));
+
+            m_BlackboardRowView = new SGBlackboardRow(m_BlackboardPropertyView, null);
+            m_BlackboardRowView.expanded = SessionState.GetBool($"Unity.ShaderGraph.Input.{shaderInput.objectId}.isExpanded", false);
+        }
+
+        void InitializeViewModel()
+        {
+            ViewModel.Model = Model;
+            ViewModel.IsSubGraph = DataStore.State.isSubGraph;
+            ViewModel.IsInputExposed = (DataStore.State.isSubGraph || (Model.isExposable && Model.generatePropertyBlock));
+            ViewModel.InputName = Model.displayName;
+            switch (Model)
+            {
+                case AbstractShaderProperty shaderProperty:
+                    ViewModel.InputTypeName = shaderProperty.GetPropertyTypeString();
+                    // HACK: Handles upgrade fix for deprecated old Color property
```

Comment:
I don't like prefixing with "HACK" - either it's a TODO we'll fix soon, or just remove the word hack and let the code speak for itself.